src/Analysis/Ast/Impl/Types/PythonClassType.cs

```diff
@@ -57,6 +53,16 @@ internal PythonClassType(string name, Location location)
             location.Module.AddAstNode(this, classDefinition);
         }
 
+        public override string Name {
+            get {
+                if (GenericParameters.Values.Count() > 0) {
```

Comment:
Although analysis of a particular module is guaranteed to run on the same task, reading data such as class name can come from multiple threads. Consider module A imported by B and C. Analysis of A is a single task, but analysis of B and C will run in parallel reading data from A. So you need some lock here. In fact, I added lock for `CreateSpecificType` since I did hit the case when two threads attempted to modify collection concurrently.

Reply:
I locked the caching part.